File: A1/include/alphapass.cpp
#include "alphapass.h"

Matrix alphapass(Matrix A, Matrix B, Matrix pi, Matrix emissionSequence, int T) {
        Matrix piT = pi.transpose();
        Matrix BColumn = B.getColumn((int) emissionSequence.get(0, 0));
        Matrix alpha = BColumn/piT;
        
        for(int t = 1; t < T; t++) {
            BColumn = B.getColumn((int) emissionSequence.get(t, 0));
            Matrix AT = A.transpose();
            alpha = AT*alpha;
            alpha = alpha/BColumn;
        }

        return alpha;
}

File: A1/HMM1_cpp/matlab/alphapass.m
function alphas = alphapass(A, B, pi, emissionSequence, T, alphas)

alpha = B(:, emissionSequence(1) + 1).*pi';

alphas = cell(T, 1);
alphas{1} = alpha;

for t=2:T
    alpha =  A'*alpha.*B(:, emissionSequence(t) + 1);
    alphas{t} = alpha;
end

end